### PR TITLE
Easter + related holidays

### PR DESCRIPTION
**Pull request by wlifuture**
Which are based on lunar calendar for Christian holidays or special days

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishHolidayParserConfiguration.cs [hunk 1]
```diff
@@ -145,6 +161,238 @@ public override string SanitizeHolidayToken(string holiday)
...
+        private static DateObject WhiteMonday(int year) => EasterDay(year).AddDays(50);
+
+        private static DateObject TrinitySunday(int year) => EasterDay(year).AddDays(56);
+
+        private static DateObject CorpusChristi(int year) => EasterDay(year).AddDays(60);
+
+        private static IList<int> GetEasterYearList() => GetEasterDateList().Select(d => d.Year).ToList();
+
+        // easter sunday: 1900 ~ 2099, following the Paschal Full Moon (PFM) date of lunar year calendar.
+        private static IList<DateObject> GetEasterDateList()
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishHolidayParserConfiguration.cs]: Doesn't German has the lunar year calculation stuff? If so, was there an issue re-using it?
If not, can't this be in the YAML file somehow?

File: Patterns/English/English-DateTime.yaml [hunk 2]
```diff
@@ -494,13 +494,13 @@ HalfRegex: !simpleRegex
...
+  def: \b(?<holiday>clean monday|good friday|ash wednesday|mardi gras|washington's birthday|mao's birthday|chinese new year|new years' eve|new year's eve|new year 's eve|new years eve|new year eve|new years'|new year's|new year 's|new years|new year|may\s*day|yuan dan|april fools|christmas eve|christmas|xmas|thanksgiving|black friday|cyber monday|halloween|yuandan|easter sunday|ash wednesday|palm sunday|maundy thursday|good friday|easter saturday|easter monday|white sunday|pentecost|white monday|trinity sunday|corpus christi|easter)(\s+(of\s+)?({YearRegex}|{RelativeRegex}\s+year))?\b
   references: [ YearRegex, RelativeRegex ]
 HolidayRegex2: !nestedRegex
-  def: \b(?<holiday>all saint's|tree planting day|white lover|st patrick|st george|cinco de mayo|us independence|all hallow|all souls|guy fawkes)(\s+(of\s+)?({YearRegex}|{RelativeRegex}\s+year))?\b
+  def: \b(?<holiday>all saint's|tree planting day|white lover|s(?:ain)?t?. patrick(?:')?(?:s)?|st george|cinco de mayo|us independence|all hallow|all souls|guy fawkes)(\s+(of\s+)?({YearRegex}|{RelativeRegex}\s+year))?\b
```
Comment (tellarin) [inline Patterns/English/English-DateTime.yaml]: Can you rebase this on /master to make it easier to review only this PR changes?

File: Specs/DateTime/English/MergedParser.json [hunk 3]
```diff
@@ -200,25 +200,26 @@
     ]
   },
   {
-    "Input": "Set an appointment for Easter",
+    "Input": "Set an appointment for Easter 2019",
```
Comment (tellarin) [inline Specs/DateTime/English/MergedParser.json]: Can you add this as a separate case? We still want to use the one without a specified year as a regression case.

File: Specs/DateTime/English/MergedParser.json [hunk 4]
```diff
@@ -793,45 +794,26 @@
     ]
   },
   {
-    "Input": "Set an appointment for Easter Sunday",
+    "Input": "Set an appointment for Easter Monday 2017",
```
Comment (tellarin) [inline Specs/DateTime/English/MergedParser.json]: It seems this case behaves different between the VSO build (it passes) and the AppVeyor one (it fails). We'll debug why.

File: Specs/DateTime/English/MergedParser.json [hunk 5]
```diff
@@ -793,45 +794,26 @@
     ]
   },
   {
-    "Input": "Set an appointment for Easter Sunday",
```
Comment (tellarin) [inline Specs/DateTime/English/MergedParser.json]: Please don't change the previous case. Add the new one as a separate test.

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishHolidayParserConfiguration.cs (hunk 1 above)
Comment (wlifuture) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishHolidayParserConfiguration.cs]: Could you point me to the Lunar implementation in German? Thanks

File: Specs/DateTime/English/MergedParser.json (hunk 5 above)
Comment (wlifuture) [reply Specs/DateTime/English/MergedParser.json]: Got it, will update shortly.

File: Specs/DateTime/English/MergedParser.json (hunk 3 above)
Comment (wlifuture) [reply Specs/DateTime/English/MergedParser.json]: Got it, will update shortly.

File: Patterns/English/English-DateTime.yaml (hunk 2 above)
Comment (wlifuture) [reply Patterns/English/English-DateTime.yaml]: Tried but seems not working for some reason, could you help to take a look? Thanks

File: .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishHolidayParserConfiguration.cs (hunk 1 above)
Comment (wlifuture) [reply .NET/Microsoft.Recognizers.Text.DateTime/English/Parsers/EnglishHolidayParserConfiguration.cs]: Thanks for pointing the German implementation. Tried to generalize the 'easter' family into base class, but got conflicts from other languages e.g. Dutch, so stepped back and just adopted the function from German into English for now.

File: Patterns/English/English-DateTime.yaml (hunk 2 above)
Comment (wlifuture) [reply Patterns/English/English-DateTime.yaml]: As discussed, ignore for now.

File: Specs/DateTime/English/MergedParser.json (hunk 3 above)
Comment (wlifuture) [reply Specs/DateTime/English/MergedParser.json]: Done

File: Specs/DateTime/English/MergedParser.json (hunk 5 above)
Comment (wlifuture) [reply Specs/DateTime/English/MergedParser.json]: Done

File: Specs/DateTime/English/MergedParser.json (hunk 4 above)
Comment (wlifuture) [reply Specs/DateTime/English/MergedParser.json]: Please take a look. With latest commit, my local build passed but PR build failed once more. Thx

**Comment by tellarin** [comment]
Close/re-open to re-trigger build.

**Pull request merged**